### EpiMethylTag simultaneously detects ATAC-seq or ChIP-seq signals with DNA methylation  
Priscillia Lhoumaud<sup>1</sup>, Gunjan Sethia<sup>1</sup>, Franco Izzo<sup>2,3</sup>, Sana Badri<sup>1</sup>, MacIntosh Cornwell<sup>1</sup>, Kyu-Tae Kim<sup>2,3</sup>, Dan Avi Landau<sup>2,3,4</sup>, Jane Skok<sup>1#</sup>.  

Affiliations:  
<sup>1</sup>New York University Langone Health, New York, NY, USA; <sup>2</sup>New York Genome Center, New York, NY, USA; <sup>3</sup>Meyer Cancer Center, Weill Cornell Medicine, New York, NY, USA; 4Institute of Computational Biomedicine, Weill Cornell Medicine, New York, NY, USA.

Correspondence: Jane.Skok@nyulangone.org

Abstract:  
Activation of regulatory elements is thought to be inversely correlated with DNA methylation levels. However, it is difficult to determine whether DNA methylation is compatible with chromatin accessibility or transcription factor (TF) binding if assays are performed separately. We developed a low input, low sequencing depth method, EpiMethylTag that combines ATAC-seq or ChIP-seq (M-ATAC or M-ChIP) with bisulfite conversion, to simultaneously examine accessibility/TF binding and methylation on the same DNA.
***
This repository contains all scripts used in computational analysis of EpiMethylTag paper.
